proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,727 @@
+  apply (clarsimp simp: find_free_hw_asid_def find_pd_for_asid_assert_def)
+  apply (intro hoare_seq_ext[OF _ gets_inv])
+  apply (clarsimp split: option.splits)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+   apply (clarsimp simp: active_cur_vcpu_of_def pred_tcb_at_def obj_at_def valid_cur_vcpu_def)
+  apply (wpsimp simp: find_pd_for_asid_assert_def)
+  done
+
+lemma arm_context_switch_active_cur_vcpu_of[wp]:
+  "arm_context_switch pd asid \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  apply (clarsimp simp: arm_context_switch_def get_hw_asid_def)
+  by (wpsimp wp: load_hw_asid_wp)
+
+lemma set_vm_root_active_cur_vcpu_of[wp]:
+  "set_vm_root tcb \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  by (wpsimp simp: set_vm_root_def find_pd_for_asid_def wp: get_cap_wp)
+
+crunches set_vm_root
+  for valid_cur_vcpu_cur_thread_update[wp]: "\<lambda>s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)"
+  (wp: valid_cur_vcpu_lift_cur_thread_update)
+
+lemma arch_switch_to_thread_valid_cur_vcpu_cur_thread_update[wp]:
+  "\<lbrace>valid_cur_vcpu\<rbrace>
+   arch_switch_to_thread t
+   \<lbrace>\<lambda>_ s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  unfolding arch_switch_to_thread_def
+  apply wpsimp
+  by (fastforce simp: active_cur_vcpu_of_def pred_tcb_at_def obj_at_def get_tcb_def
+               split: option.splits kernel_object.splits)
+
+lemma switch_to_thread_valid_cur_vcpu[wp]:
+  "switch_to_thread t \<lbrace>valid_cur_vcpu\<rbrace>"
+  unfolding switch_to_thread_def
+  by (wpsimp simp: valid_cur_vcpu_def active_cur_vcpu_of_def)
+
+lemma arch_switch_to_idle_thread_valid_cur_vcpu_cur_thread_update[wp]:
+  "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> valid_idle s \<and> t = idle_thread s\<rbrace>
+   arch_switch_to_idle_thread
+   \<lbrace>\<lambda>_ s. valid_cur_vcpu (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  unfolding arch_switch_to_idle_thread_def
+  apply wpsimp
+  by (clarsimp simp: valid_idle_def pred_tcb_at_def obj_at_def valid_arch_idle_def)
+
+lemma switch_to_idle_thread_valid_cur_vcpu[wp]:
+  "\<lbrace>valid_cur_vcpu and valid_idle\<rbrace>
+   switch_to_idle_thread
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  by (wpsimp simp: switch_to_idle_thread_def)
+
+lemma dissociate_vcpu_tcb_valid_cur_vcpu:
+  "\<lbrace>\<lambda>s. valid_cur_vcpu s \<and> sym_refs (state_hyp_refs_of s)\<rbrace>
+   dissociate_vcpu_tcb vr t
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  apply (clarsimp simp: dissociate_vcpu_tcb_def)
+  apply (clarsimp simp: arch_thread_get_def)
+  apply (rule hoare_seq_ext[OF _ gets_the_sp])
+  apply (clarsimp simp: get_vcpu_def bind_assoc)
+  apply (rule hoare_seq_ext[OF _ get_object_sp])
+  apply (clarsimp split: kernel_object.splits arch_kernel_obj.splits)
+  apply (rule hoare_seq_ext[OF _ assert_sp])
+  apply (rule hoare_seq_ext[OF _ gets_sp])
+  apply (simp flip: bind_assoc)
+  apply (rule hoare_seq_ext)
+   apply wpsimp
+  apply (simp add: bind_assoc)
```

Comment:
This looks like too much explicit forward single-stepping to me for a proof like this. Those single steps are very brittle in maintenance, because they spell out each single step of the function in the proof, so even small reorderings/additions will break. They shouldn't really be necessary here. Of course, that's hard to say from just the diff. Some of that could be because of the inline unfolding of `vcpu_invalidate_active`, which probably needs its own lemma(s).

The lemma below `associate_vcpu_tcb_valid_cur_vcpu` has the same flavour.`wp`/`wpsimp` should be effective on these, but they probably need some hand holding in the form of lemmas.